### PR TITLE
cmd/ensdns: Initial implementation of an ENS<->DNS gateway

### PR DESCRIPTION
**Pull request by Arachnid**

### CODE REVIEW

**Comment by fjl** [comment]
Not sure why you're adding another contract binding. We already have an ENS binding in contracts/ens.

**Comment by fjl** [comment]
You should add "github.com/miekg/dns" to vendor/ directory.

**Comment by Arachnid** [comment]
PTAL.

File: cmd/ensdns/main.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+//go:generate abigen --sol contract/ens.sol --pkg contract --out contract/ens.go
+//go:generate abigen --sol contract/resolver.sol --pkg contract --out contract/resolver.go
```
Comment (fjl) [inline cmd/ensdns/main.go]: Please remove these lines.

File: cmd/ensdns/main.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/contracts/ens/contract"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/miekg/dns"
+)
+
+var (
+	gethAddressFlag     = flag.String("geth", "ws://localhost:8546", "Path to connect to geth on")
```
Comment (fjl) [inline cmd/ensdns/main.go]: Please call this "ethapi", like we do for cmd/swarm.

File: cmd/ensdns/main.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	registryAddressFlag = flag.String("registry", "", "Address of ENS registry")
+	listenAddressFlag   = flag.String("address", ":8053", "Address and port to listen on")
```
Comment (fjl) [inline cmd/ensdns/main.go]: Shouldn't it be :53 by default?

File: cmd/ensdns/main.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+		log.Fatalf("Error constructing ENSDNS: %v", err)
+	}
+
+	dns.HandleFunc(".", ensdns.Handle)
+	go serve(*listenAddressFlag)
+
+	log.Printf("Listening on %s", *listenAddressFlag)
+
+	sig := make(chan os.Signal)
+	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
```
Comment (fjl) [inline cmd/ensdns/main.go]: You don't need to handle signals like this. The default behaviour is to quit the program on interrupt.

File: cmd/ensdns/main.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+		log.Fatalf("Error connecting to geth: %v", err)
+	}
+
+	ensdns, err := New(client, common.HexToAddress(*registryAddressFlag))
+	if err != nil {
+		log.Fatalf("Error constructing ENSDNS: %v", err)
+	}
+
+	dns.HandleFunc(".", ensdns.Handle)
+	go serve(*listenAddressFlag)
```
Comment (fjl) [inline cmd/ensdns/main.go]: Don't listen in a goroutine. Just call ListenAndServe on the main goroutine.

File: cmd/ensdns/main.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+		log.Fatalf("DNS server failed: %v", err)
+	}
+}
+
+func main() {
+	flag.Parse()
+
+	client, err := ethclient.Dial(*gethAddressFlag)
+	if err != nil {
+		log.Fatalf("Error connecting to geth: %v", err)
```
Comment (fjl) [inline cmd/ensdns/main.go]: "connecting to geth" -> "connecting to the Ethereum API"

**Comment by fjl** [comment]
One more thing: please use GPL headers instead of LGPL source file headers because source in cmd/ is GPL.

File: cmd/ensdns/main.go (hunk 5 above)
Comment (Arachnid) [inline cmd/ensdns/main.go]: What if I later need to start more than one daemon?

File: cmd/ensdns/main.go (hunk 4 above)
Comment (Arachnid) [inline cmd/ensdns/main.go]: Right, but we want to wait until we get an interrupt before halting.

Comment (fjl) [inline cmd/ensdns/main.go]: ???

Comment (fjl) [inline cmd/ensdns/main.go]: You can just put `select{}` if you don't want to run the listener on the main goroutine.

**Comment by Arachnid** [comment]
PTAL

**Comment by fjl** [comment]
We have decided to move ensdns into its own repo for now.

**Pull request closed without merging**